### DISCUSSION TITLE
Network Device Driver support

### DISCUSSION THREAD

**Discussion by keouchao** [OpenAMP/open-amp#264] [General]
Our project using zcu102 like architecture with R5 control peripheral without OS, A53 running linux. The device communicate with each other directly over R5 with a propitiatory interface. 
The idea is to add a Network device driver over share memory (openamp), so each device can communicate with each other via standard IP connections. 
Like to know does anyone has implemented something alike. Thanks.

**Comment by supergaute** [comment]
I've implemented an RPMsg network (layer 3) interface for lwip, which goes to a TUN interface on the Linux side.
I then configured Linux netfilters to forward data between physical network and the OpenAMP application.

I started to write a Virtio network driver for lwip which would be used together with OpenAMP (which would probably have been a better solution) but abandoned that idea because the incomplete Virtio support in OpenAMP

**Reply by keouchao** [reply]
Thanks for the tips. 
You said layer 3 RSMsg network, is it from your requirements or there is a limitation that no layer 2 can be supported ? since we need layer 2 over OpenAMP interface. Also, is your lwip implementation in userland ?

**Reply by supergaute** [reply]
The reason why I went for layer 3 (IP) instead of layer 2 (Ethernet) is that this is a point-to-point link, and it is not necessary to add an ethernet layer to my application.
It will also work with layer 2 data (with ARP in lwip) and a TAP interface in Linux, if you need this.
Since I run lwip with FreeRTOS/Bare metal, there is no real userspace/kernelspace separation.

**Reply by keouchao** [reply]
I misunderstood your earlier reply. Is the flow like this
apps <-> lwip <-> RSMsg @ R5 (FreeRTOS) <-> RSMsg @ A53 (Linux) <-> TUN <-> netfilters <-> ...
If my understanding is correct, can you share more details on the Linux side between RSMsg and TUN.
Thanks,

**Reply by supergaute** [reply]
Yes, that is correct flow.
The Linux side is fairly simple, a userspace application which copies data between an lwip RPMsg endpoint and the TUN interface.
